### ISSUE TITLE
DDS Topology: Add support for task assets

### ISSUE THREAD

**Issue by AnarManafov** [FairRootGroup/DDS#406]
Background info: https://alice.its.cern.ch/jira/browse/R3C-696

DDS Topology should give a possibility to specify assets, which then can be used by tasks.
- An asset is represented as a file. 
- Each asset should be described with a full path.
- The path to the asset is either on the local to DDS commander system or a local to DDS agents hosts. This can be indicated by the "reachable" flag.
- If the given tasks uses the asset, then the asset should be copied to the corresponding task slot before the task is started. A task slot directory is a working directory for the tasks.
- One asset can be assigned to multiple tasks.

**Comment by davidrohr** [comment]
@AnarManafov : Thx for this feature. I think your description lacks one aspect that is important for us:
We would need that the asset can be part of the XML, i.e. it is not provided by an external file, but it is embedded in the DDS XML topology. The reason is the way we create the topologies:
- ODC executes a script, that outputs the DDS XML topology to use to stdout.
- This script runs the PDP workflow generation, and it creates the DDS XML topology, but also some JSON files that are needed by the tasks.
- From there on ODC takes over the DDS XML file, our workflow machinery has no knowledge how long the XML file is used. (Also, detector experts might run our machinery instead to create an XML file, which they then pass to DDS).

Currently, we have to cache all our config files locally, but since we do not know the lifetime of the corresponding XML file, we have to store them indefinitely. I would very much prefer to couple these config files (which are anyway tight to the DDS XML) inside the XML file. In that way, they are automatically available as soon as someone has the XML, and when the XML is deleted, they are automatically cleaned up.

My example in the ticket you linked as background info gives an example how that could be done.

Also, the multiple assets per tasks might become relevant for us soon, so I would prefer not to move it too far into the future.

**Comment by AnarManafov** [comment]
@davidrohr , thanks! It's a fair point and makes sense.
I will implement two types of assets: <task_asset> and <task_inline_asset>

The first one is an external files and can be ascii or binary. The last one is the inlined file - can be only ascii. 
It's possible to do inline as a binary, but it will require a conversion, like a script with a binary payload.
But note, the user will be responsible to escape XML special symbols in the inline assets. 
After the release of the feature I will also think on how to make the escaping implemented automatically by DDS. Should be doable.

**Comment by davidrohr** [comment]
Hi @AnarManafov : If I understand correctly, I would use them in the command line. I.e. I could do something like (using proper escaping)
```
<exe reachable="true">echo ${configFileContent} > config.cfg; runProgram</exe>
```
or in our case
```
<exe reachable="true">echo ${configFileContent} | o2-dpl-run</exe>
```
I remember we discussed this, and concluded it could lead to issues like too long command line, etc.
Note that the config file size can be > 100 kb.

Can you tell me what order of magnitude is the ETA for having assets? If that is 2 months, I would just rather wait instead of considering alternatives. If it is a year, ok, then we have to take action.

**Comment by AnarManafov** [comment]
Hey @davidrohr,

with a 100KB I would also agree, that having assets is a better solution. 
As for the ETA. Yes, around that time. I am working on the design for it in parallel to other tickets. 
During the design stage I am also looking for edge cases. For example, I learned that assets should be hard linked with their tasks. I mean in the agent's code. If a tasks exists or crashes, then all assets must be removed for this task by the agent. Otherwise consistency will be broken and tasks might get assets of prevues tasks, etc. Basically looking for such things, which must be accounted in development.

**Comment by AnarManafov** [comment]
Hey @davidrohr,

with a 100KB I would also agree, that having assets is a better solution. 
As for the ETA. Yes, around that time. I am working on the design for it in parallel to other tickets. 
During the design stage I am also looking for edge cases. For example, I learned that assets should be hard linked with their tasks. I mean in the agent's code. If a tasks exists or crashes, then all assets must be removed for this task by the agent. Otherwise consistency will be broken and tasks might get assets of previous tasks, etc. Basically looking for such things, which must be accounted in development.

**Comment by davidrohr** [comment]
OK, then it will be no problem to wait until assets are available.
For our case, I will make sure that assets have unique names in the topology, so they should never collide (same as the `<exe...> --id [IDS]`).

**Comment by davidrohr** [comment]
@AnarManafov : What is the status of this feature? In Januarry we aimed before end of March. Is this still realistic?

**Comment by AnarManafov** [comment]
@davidrohr , at the moment I am working with Alexey on the support for minimum number of tasks logic. After that I will be able to start working on it, if nothing important comes up.

**Comment by davidrohr** [comment]
@AnarManafov : Could you update us on the ETA when this feature will be available?

**Comment by AnarManafov** [comment]
Hey @davidrohr, I am working on it and on #414. If nothing else urgent comes up, then sometime next week it should be in the master.

**Comment by AnarManafov** [comment]
@davidrohr , the feature is ready. It passed internal tests and is merged into the master now.
I also tagged the latest DDS as "3.7.7".

Please, give it a try. Feel free to reopen the ticket if something is wrong or missing.
I didn't yet have a chance to write a doc, therefore wanted to give you a couple of details here so that you can try rightaway.

In the topology file you can now define task assets:

~~~~
<asset name="my_asset_name" type="inline" value="THE CONTENT OF THE ASSET"/>
...
...
  	<decltask name="task1">
		<exe reachable="true">sleep 180</exe>
	    <assets>
				<name>my_asset_name</name>
				<name>asset2</name>
				<name>asset3</name>
		</assets>
	</decltask>
~~~~

Please note, the value must be escaped for XML special symbols.
When serializing the asset into the file, the escpaed symbols will be replaced to real values.
I.e. a symbol "<" must be escaped as "&lt;". But in the desitnation file it will be saved as "<".
This applies for all XML special symbols. 

* The size of the value of the asset is not limited. Probably I would need to put some reasonable limits, if users start to overuse it. But for the moment there are no limits.
* Assets are saved to the Task's working directory as "<assetName>_<taskName>.asset" before the task is executed.
* Each task can have multiple assets. Each asset will be saved in a seporate file, of course.
* All assets of the given task are automaticlly removed after the task is exited.

CC: @rbx

**Comment by davidrohr** [comment]
@AnarManafov : Thx, I'll give it a try asap. I have few questions:
- Do I understand correctly, that when I have an asset with `<name>my_asset_name</name>`, the filename fill be `my_asset_name.asset`?
- What characters do we have to escape? Since it is in a double quote, why do we have to escape `<`? I had assumed we only need to escape double quotes and backslashes?
- What happens if multiple tasks use the same assets? Actually, do the tasks share the same working directory when they are in one collection? If so, will they then all access the same asset file? That would make sense in my opinion to avoid asset file duplication.

**Comment by AnarManafov** [comment]
@davidrohr, let me address your questions.

1. Not exactly. There will be also a task name, i.e. "my_asset_name_taskName.asset". Sorry, I didn't noticed that in my initial comment the name of the file was escaped. It's fixed now
2. I am afraid, if this is not escaped, then the xml will not pass the validation.  At least, that was the case with "<". I just used to escape all the reset too. To be on the safe side.
3. In DDS all tasks are isolated and run in dedicated slot directories. Same collection only means that they run under control of one DDS agent (one host), but all the reset is an isolated environment per task. So, if multiple tasks share the same asset, the asset will be duplicated for each task in their working dirs. Isolated environment is important as not all tasks run exactly the same amount of time. Some might fail, some might be shutdown and replaced on a topology update command.

Ping me if you need more info.

**Comment by davidrohr** [comment]
@AnarManafov : Thx for the clarification.

I am a bit confused about the working directory. I checked the stderr Monitor tool we are using, and this is also a task started by DDS. It opens the log files in folder `.`, which should be its working directory, and this is something like `/tmp/wn_dds/02076219-6877-442e-99dc-90f7cf693896/wn_online_epn002-ib.internal`, so that seems to be common for all tasks.
Inside it, I find a folder slots, and inside it the 384 subdirectories, one per slot. But to me it seems the working directory is not inside slot/***, but it is in the directory below, the same where all the log files are written to.

The fact that the task name goes into the asset file might make it a bit complicated. Is the task name available as a variable, that I can use in the `<exe` tag. I.e. let's assume I have the asset `my_asset_name`, and I want to pass it to a task on the command line argument, e.g. 
```<exe reachable="true">some_binary --configFile my_asset_name.asset</exe>```
How can I do this if the asset is ammended by the task name?

**Comment by AnarManafov** [comment]
Hey @davidrohr ,

reg. name of the asset file. Honestly I had something in mind when I added the task name. But I can't recall it now. So I guess you are right, having file named as "assetName.asset" is enough.
BTW, just for info, DDS exports a set of environment variables for each task. One of which is $DDS_TASK_NAME so a task can easily get its topology name. Other env vars are: DDS_TASK_ID, DDS_TASK_INDEX, DDS_COLLECTION_INDEX, DDS_TASK_PATH, DDS_GROUP_NAME, DDS_COLLECTION_NAME, DDS_TASK_NAME, DDS_SLOT_ID
Task can use those to get a better understanding of what is going on :)

reg. the working directory.
The logs of tasks are located in $DDS_LOCATION folder, that what the mon tool should use. But the working dir should point to DDS_TASK_PATH. If it is not the case, then it's a bug. 

Give me a moment to fix the name of the asset and check the tasks working dir.

**Comment by AnarManafov** [comment]
@davidrohr , I have fixed the name of the assets and fixed the bug with the tasks working directory. The bug was a leftover after we improved DDS to use one agent for multiple tasks.

* Assets are now "assetName.asset" and located in the task working directory.
* Tasks working directory is pointing to their slot directory.
* The logs of tasks are still located in $DDS_LOCATION of agents (WNs). This helps to safely monitor and collect them. But you would need to adjust your monitoring tool to use $DDS_LOCATION instead of the working directory.

A new developer release is tagged as "3.7.8"

**Comment by davidrohr** [comment]
Hi Anar,

OK, I am a bit concerned about the change of the working directory. We need to test this before we can use it. SInce I am not sure if something else relies on it. But OK, I assume this change should be done in any case, so I'll have a look.

For the asset files: One point that comes to my mind: We have a ~1 MB config file. With 384 tasks per node, that is almost have a gigabyte of config files. The files are removed when the tasks are done you said?
Perhaps, we could apply an optimization and put the asset file in the DDS folder, and then just symlink it into the task folder. Or even the symlink would not be necessary, I could just use `$DDS_LOCATION/my_asset_name.asset` ? What do you think?

**Comment by AnarManafov** [comment]
Regarding the working dir. Yeah. I am afraid that was a bug and it needs to be fixed. For the time when we had one agent per task, it was fine to have working dir in the DDS_LOCATION. But now tasks are sandboxed per slot. 

Regarding the assets location. I agree with you, it makes perfect sense.
I still want to have assets also isolated per tasks. Therefore I will introduce a "visibility" attribute. It can be "local" (the asset will be placed in the slot dir) and "global" (the asset will be placed in DDS_LOCATION).
* Local assets are created before starting a task and removed when the task is done.
* Global assets are created before tasks are started, but will be removed only on session end.

How does it sound?

**Comment by AnarManafov** [comment]
@davidrohr, I introduced a new attribute - asset visibility. It can have one of the two values at the moment: "task" or "global".
Task assets are saved to the tasks working dir (their slot dir).
Global assets are saved into $DDS_LOCATION of the agent.

Please note, the global assets are not updated on topology update. Once created, they will stay unchanged until session shutdown.

Usage:
~~~
<asset name="asset1" type="inline" visibility="task" value="...."/>
<asset name="asset2" type="inline" visibility="global" value="...."/>

<decltask name="task1">
		<exe reachable="true">sleep 180</exe>
	    <assets>
				<name>asset1</name>
				<name>asset2</name>
		</assets>
	</decltask>
~~~

DDS tag: 3.7.9

Give it a try and let me know if you have comments.

**Comment by davidrohr** [comment]
@AnarManafov : Do I need to specify global assets in the `decltask`? Won't they be available in any case?

**Comment by AnarManafov** [comment]
@davidrohr , yes, as is in my example above. 
A global asset is not global for the topology, it’s global for an agent of the dependent task. 
I would like to have tasks as the main source of dependencies.
